tests/raftstore/test_stale_peer.rs

```diff
@@ -0,0 +1,179 @@
+use kvproto::raft_serverpb::{RegionLocalState, PeerState};
+use tikv::raftstore::store::{keys, Peekable};
+
+use super::cluster::{Cluster, Simulator};
+use super::node::new_node_cluster;
+use super::server::new_server_cluster;
+use super::transport_simulate::{Isolate, IsolateRegionStore};
+use super::util::*;
+
+/// A helper function for testing the behaviour of the gc of stale peer
+/// which is out of region.
+/// If a peer detects the leader is missing for a specified long time,
+/// it should consider itself as a stale peer which is removed from the region.
+/// This test case covers the following scenario:
+/// At first, there are three peer A, B, C in the cluster, and A is leader.
+/// Peer B gets down. And then A adds D, E, F int the cluster.
+/// Peer D becomes leader of the new cluster, and then removes peer A, B, C.
+/// After all these peer in and out, now the cluster has peer D, E, F.
+/// If peer B goes up at this moment, it still thinks it is one of the cluster
+/// and has peers A, C. However, it could not reach A, C since they are removed from
+/// the cluster or probably destroyed.
+/// Meantime, D, E, F would not reach B, Since it's not in the cluster anymore.
+/// In this case, Peer B would notice that the leader is missing for a long time,
+/// and it would check with pd to confirm whether it's still a member of the cluster.
+/// If not, it should destroy itself as a stale peer which is removed out already.
+fn test_stale_peer_out_of_region<T: Simulator>(cluster: &mut Cluster<T>) {
+    // Revise raft base tick interval small to make this test case run fast.
+    cluster.cfg.raft_store.raft_base_tick_interval = 10;
+    // Specify the heartbeat and election timeout ticks here
+    // so that this test case works without depending on the default value of them
+    cluster.cfg.raft_store.raft_heartbeat_ticks = 3;
+    cluster.cfg.raft_store.raft_election_timeout_ticks = 15;
+    // Use a value of 3 seconds as max time here just for test.
+    // In production environment, the value of max_leader_missing_duration
+    // should be configured far beyond the election timeout.
+    let max_leader_missing_duration = Duration::from_secs(3);
+    cluster.cfg.raft_store.max_leader_missing_duration = max_leader_missing_duration;
+    let pd_client = cluster.pd_client.clone();
+    // Disable default max peer number check.
+    pd_client.disable_default_rule();
+
+    let r1 = cluster.run_conf_change();
+    pd_client.must_add_peer(r1, new_peer(2, 2));
+    pd_client.must_add_peer(r1, new_peer(3, 3));
+    let (key, value) = (b"k1", b"v1");
+    cluster.must_put(key, value);
+    assert_eq!(cluster.get(key), Some(value.to_vec()));
+
+    let engine_2 = cluster.get_engine(2);
+    must_get_equal(&engine_2, key, value);
+
+    // isolate peer 2 from other part of the cluster
+    cluster.add_send_filter(Isolate::new(2));
+
+    // add peer [(4, 4), (5, 5), (6, 6)]
+    pd_client.must_add_peer(r1, new_peer(4, 4));
```

Comment:
> 1 adds 4, 5, 6, Now assume 4 is leader, and 4 removes 1, 3 first, then remove 2. 
> When 2 re-joins the cluster, it will send vote to 1, 3, ...

@BusyJay The "remove" is ahead of "re-join".